### DISCUSSION TITLE
[SWIP-8] Support Kong monitoring

### DISCUSSION THREAD

**Discussion by CodePrometheus** [apache/skywalking#12629] [SWIP]
# Support Kong Monitoring

## Motivation

[**Kong**](https://github.com/Kong/kong) or **Kong API Gateway** is a cloud-native, platform-agnostic, scalable API Gateway 
distinguished for its high performance and extensibility via plugins. Now I want to add Kong monitoring via the OpenTelemetry Collector, 
which fetches metrics from it's own HTTP endpoint to expose metrics data for [Prometheus](https://prometheus.io/).

## Architecture Graph

There is no significant architecture-level change.

## Proposed Changes

1. Kong expose its own [metrics](https://docs.konghq.com/hub/kong-inc/prometheus/) via HTTP endpoint to opentelemetry collector, OpenTelemetry Collector fetches metrics from it and pushes metrics to SkyWalking OTEL Receiver via OpenTelemetry exporter.
2. The SkyWalking OAP Server parses the expression with MAL to filter/calculate/aggregate and store the results.
3. These metrics can be displayed via the SkyWalking UI, and the metrics can be customized for display on the UI dashboard.

### Kong Request Supported Metrics

| Monitoring Panel | Unit  | Metric Name                                                                                                       | Description                                          | Data Source |
|------------------|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------|-------------|
| Bandwidth        | bytes | meter_kong_service_http_bandwidth<br />meter_kong_instance_http_bandwidth<br />meter_kong_endpoint_http_bandwidth | Total bandwidth (ingress/egress) throughput          | Kong        |
| HTTP Status      | count | meter_kong_service_http_status<br />meter_kong_instance_http_status<br />meter_kong_endpoint_http_status          | HTTP status codes per consumer/service/route in Kong | Kong        |
| HTTP Request     | count | meter_kong_service_http_requests<br />meter_kong_instance_http_requests                                           | Total number of requests                             | Kong        |

### Kong Database Supported Metrics

| Monitoring Panel | Unit  | Metric Name                                                                         | Description                               | Data Source |
|------------------|-------|-------------------------------------------------------------------------------------|-------------------------------------------|-------------|
| DB               | count | meter_kong_service_datastore_reachable<br />meter_kong_instance_datastore_reachable | Datastore reachable from Kong             | Kong        |
| DB               | bytes | meter_kong_instance_shared_dict_bytes                                               | Allocated slabs in bytes in a shared_dict | Kong        |
| DB               | bytes | meter_kong_instance_shared_dict_total_bytes                                         | Total capacity in bytes of a shared_dict  | Kong        |
| DB               | bytes | meter_kong_instance_memory_workers_lua_vms_bytes                                    | Allocated bytes in worker Lua VM          | Kong        |

### Kong Latencies Supported Metrics

| Monitoring Panel | Unit | Metric Name                                                                                                             | Description                                                              | Data Source |
|------------------|------|-------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------|-------------|
| Latency          | ms   | meter_kong_service_kong_latency<br />meter_kong_instance_kong_latency<br />meter_kong_endpoint_kong_latency             | Latency added by Kong and enabled plugins for each service/route in Kong | Kong        |
| Latency          | ms   | meter_kong_service_request_latency<br />meter_kong_instance_request_latency<br />meter_kong_endpoint_request_latency    | Total latency incurred during requests for each service/route in Kong    | Kong        |
| Latency          | ms   | meter_kong_service_upstream_latency<br />meter_kong_instance_upstream_latency<br />meter_kong_endpoint_upstream_latency | Latency added by upstream response for each service/route in Kong        | Kong        |

### Kong Nginx Supported Metrics

| Monitoring Panel | Unit  | Metric Name                                                                                 | Description                           | Data Source |
|------------------|-------|---------------------------------------------------------------------------------------------|---------------------------------------|-------------|
| Nginx            | count | meter_kong_service_nginx_metric_errors_total                                                | Number of nginx-lua-prometheus errors | Kong        |
| Nginx            | count | meter_kong_service_nginx_connections_total<br />meter_kong_instance_nginx_connections_total | Number of connections by subsystem    | Kong        |
| Nginx            | count | meter_kong_service_nginx_timers<br />meter_kong_instance_nginx_timers                       | Number of Nginx timers                | Kong        |

## Imported Dependencies libs and their licenses.

No new dependency.

## Compatibility

no breaking changes.

**Comment by wu-sheng** [comment]
I think we are good to have this.

**Reply by wu-sheng** [reply]
You could update the next swip ID first, and merge the swip-8 doc later with the feature.

**Comment by wu-sheng** [comment]
@CodePrometheus We are going to release 10.1 just before the National Day or after that holiday. Could you share are you going to add this into 10.1 or later?

**Reply by CodePrometheus** [reply]
I really hope it will be 10.1, but at the current pace, it's highly likely that I won't be able to catch up before the holiday.

**Reply by wu-sheng** [reply]
Thanks for the update. We will see whether we are possible to get there. There is one more feature of BanyanDB 0.7 integration.

**Comment by CodePrometheus** [comment]
![image](https://github.com/user-attachments/assets/72c1b234-00e9-4372-aa79-7cbacfeaa391)

As the feature progressed, I find that the metrics in the first version could has a better way of handling and categorized, I will make some adjustments to the SWIP and submit a PR when I'm ready.

**Reply by wu-sheng** [reply]
OK. Take your time. I will initial the 10.1 release before Oct 1st, I think. You could put this in 10.2.

**Reply by wu-sheng** [reply]
From what I saw from your screenshot, I would like to ask, would you have route rule metrics? I could set them as endpoint.

**Reply by CodePrometheus** [reply]
Yes, I divide them into service (cluster), instance and endpoint (route).
